// @sap/cds-dk/lib/init/template/sample/files/@cds-models/AdminService/index.ts
// This is an automatically generated file. Please do not change its contents manually!
import * as _ from './..';
import * as __ from './../_';
import * as _sap_capire_bookshop from './../sap/capire/bookshop';
import * as _sap_common from './../sap/common';

export default class {
}

export function _AuthorAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Author extends Base {
    declare createdAt?: __.CdsTimestamp | null
    /** Canonical user ID */
    declare createdBy?: _.User | null
    declare modifiedAt?: __.CdsTimestamp | null
    /** Canonical user ID */
    declare modifiedBy?: _.User | null
    declare ID?: __.Key<number>
    declare name?: string | null
    declare dateOfBirth?: __.CdsDate | null
    declare dateOfDeath?: __.CdsDate | null
    declare placeOfBirth?: string | null
    declare placeOfDeath?: string | null
    declare books?: __.Association.to.many<Books>
    declare age?: number | null
    declare lifetime?: string | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Author>;
    declare static readonly elements: __.ElementsOf<Author>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
export class Author extends _AuthorAspect(__.Entity) {static drafts: __.DraftOf<Author>}
Object.defineProperty(Author, 'name', { value: 'AdminService.Authors' })
Object.defineProperty(Author, 'is_singular', { value: true })
export class Authors extends Array<Author> {static drafts: __.DraftsOf<Author>
$count?: number}
Object.defineProperty(Authors, 'name', { value: 'AdminService.Authors' })

export function _BookAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Book extends Base {
    declare createdAt?: __.CdsTimestamp | null
    /** Canonical user ID */
    declare createdBy?: _.User | null
    declare modifiedAt?: __.CdsTimestamp | null
    /** Canonical user ID */
    declare modifiedBy?: _.User | null
    declare ID?: __.Key<number>
    declare author?: __.Association.to<Author> | null
    declare author_ID?: number | null
    declare title?: string | null
    declare descr?: string | null
    declare genre?: __.Association.to<Genre> | null
    declare genre_ID?: string | null
    declare stock?: number | null
    declare price?: _sap_capire_bookshop.Price | null
    /**
    * Type for an association to Currencies
    * 
    * See https://cap.cloud.sap/docs/cds/common#type-currency
    */
    declare currency?: _.Currency | null
    declare currency_code?: string | null
    declare texts?: __.Composition.of.many<Books.texts>
    declare localized?: __.Association.to<Books.text> | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Book>;
    declare static readonly elements: __.ElementsOf<Book>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
export class Book extends _BookAspect(__.Entity) {static drafts: __.DraftOf<Book>}
Object.defineProperty(Book, 'name', { value: 'AdminService.Books' })
Object.defineProperty(Book, 'is_singular', { value: true })
export class Books extends Array<Book> {static drafts: __.DraftsOf<Book>
$count?: number}
Object.defineProperty(Books, 'name', { value: 'AdminService.Books' })

export function _GenreAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Genre extends Base {
    declare ID?: __.Key<string>
    declare name?: string | null
    declare descr?: string | null
    declare parent?: __.Association.to<Genre> | null
    declare parent_ID?: string | null
    declare children?: __.Composition.of.many<Genres>
    declare texts?: __.Composition.of.many<Genres.texts>
    declare localized?: __.Association.to<Genres.text> | null
    declare LimitedDescendantCount?: number | null
    declare DistanceFromRoot?: number | null
    declare DrillState?: string | null
    declare LimitedRank?: number | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Genre>;
    declare static readonly elements: __.ElementsOf<Genre>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
/** Hierarchically organized Code List for Genres */
export class Genre extends _GenreAspect(__.Entity) {}
Object.defineProperty(Genre, 'name', { value: 'AdminService.Genres' })
Object.defineProperty(Genre, 'is_singular', { value: true })
/** Hierarchically organized Code List for Genres */
export class Genres extends Array<Genre> {$count?: number}
Object.defineProperty(Genres, 'name', { value: 'AdminService.Genres' })

export function _LanguageAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Language extends Base {
    declare name?: string | null
    declare descr?: string | null
    /** Type for a language code */
    declare code?: __.Key<_sap_common.Locale>
    declare texts?: __.Composition.of.many<Languages.texts>
    declare localized?: __.Association.to<Languages.text> | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Language>;
    declare static readonly elements: __.ElementsOf<Language>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
/**
* Code list for languages
* 
* See https://cap.cloud.sap/docs/cds/common#entity-languages
*/
export class Language extends _LanguageAspect(__.Entity) {}
Object.defineProperty(Language, 'name', { value: 'AdminService.Languages' })
Object.defineProperty(Language, 'is_singular', { value: true })
/**
* Code list for languages
* 
* See https://cap.cloud.sap/docs/cds/common#entity-languages
*/
export class Languages extends Array<Language> {$count?: number}
Object.defineProperty(Languages, 'name', { value: 'AdminService.Languages' })

export function _CurrencyAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Currency extends Base {
    declare name?: string | null
    declare descr?: string | null
    declare code?: __.Key<string>
    declare symbol?: string | null
    declare minorUnit?: number | null
    declare numcode?: number | null
    declare exponent?: number | null
    declare minor?: string | null
    declare texts?: __.Composition.of.many<Currencies.texts>
    declare localized?: __.Association.to<Currencies.text> | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Currency>;
    declare static readonly elements: __.ElementsOf<Currency>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
/**
* Code list for currencies
* 
* See https://cap.cloud.sap/docs/cds/common#entity-currencies
*/
export class Currency extends _CurrencyAspect(__.Entity) {}
Object.defineProperty(Currency, 'name', { value: 'AdminService.Currencies' })
Object.defineProperty(Currency, 'is_singular', { value: true })
/**
* Code list for currencies
* 
* See https://cap.cloud.sap/docs/cds/common#entity-currencies
*/
export class Currencies extends Array<Currency> {$count?: number}
Object.defineProperty(Currencies, 'name', { value: 'AdminService.Currencies' })

export namespace Books {
  export function _textAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
    return class text extends Base {
      declare ID_texts?: __.Key<string>
      /** Type for a language code */
      declare locale?: _sap_common.Locale | null
      declare ID?: number | null
      declare title?: string | null
      declare descr?: string | null
      static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
      declare static readonly keys: __.KeysOf<text>;
      declare static readonly elements: __.ElementsOf<text>;
      declare static readonly actions: globalThis.Record<never, never>;
    };
  }
  export class text extends _textAspect(__.Entity) {static drafts: __.DraftOf<text>}
  Object.defineProperty(text, 'name', { value: 'AdminService.Books.texts' })
  Object.defineProperty(text, 'is_singular', { value: true })
  export class texts extends Array<text> {static drafts: __.DraftsOf<text>
$count?: number}
  Object.defineProperty(texts, 'name', { value: 'AdminService.Books.texts' })
  
}
export namespace Genres {
  export function _textAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
    return class text extends Base {
      /** Type for a language code */
      declare locale?: __.Key<_sap_common.Locale>
      declare ID?: __.Key<string>
      declare name?: string | null
      declare descr?: string | null
      static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
      declare static readonly keys: __.KeysOf<text>;
      declare static readonly elements: __.ElementsOf<text>;
      declare static readonly actions: globalThis.Record<never, never>;
    };
  }
  export class text extends _textAspect(__.Entity) {}
  Object.defineProperty(text, 'name', { value: 'AdminService.Genres.texts' })
  Object.defineProperty(text, 'is_singular', { value: true })
  export class texts extends Array<text> {$count?: number}
  Object.defineProperty(texts, 'name', { value: 'AdminService.Genres.texts' })
  
}
export namespace Languages {
  export function _textAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
    return class text extends Base {
      /** Type for a language code */
      declare locale?: __.Key<_sap_common.Locale>
      declare name?: string | null
      declare descr?: string | null
      /** Type for a language code */
      declare code?: __.Key<_sap_common.Locale>
      static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
      declare static readonly keys: __.KeysOf<text>;
      declare static readonly elements: __.ElementsOf<text>;
      declare static readonly actions: globalThis.Record<never, never>;
    };
  }
  export class text extends _textAspect(__.Entity) {}
  Object.defineProperty(text, 'name', { value: 'AdminService.Languages.texts' })
  Object.defineProperty(text, 'is_singular', { value: true })
  export class texts extends Array<text> {$count?: number}
  Object.defineProperty(texts, 'name', { value: 'AdminService.Languages.texts' })
  
}
export namespace Currencies {
  export function _textAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
    return class text extends Base {
      /** Type for a language code */
      declare locale?: __.Key<_sap_common.Locale>
      declare name?: string | null
      declare descr?: string | null
      declare code?: __.Key<string>
      static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
      declare static readonly keys: __.KeysOf<text>;
      declare static readonly elements: __.ElementsOf<text>;
      declare static readonly actions: globalThis.Record<never, never>;
    };
  }
  export class text extends _textAspect(__.Entity) {}
  Object.defineProperty(text, 'name', { value: 'AdminService.Currencies.texts' })
  Object.defineProperty(text, 'is_singular', { value: true })
  export class texts extends Array<text> {$count?: number}
  Object.defineProperty(texts, 'name', { value: 'AdminService.Currencies.texts' })
  
}
