// @sap/cds-dk/lib/init/template/sample/files/@cds-models/sap/capire/bookshop/index.js
// This is an automatically generated file. Please do not change its contents manually!
const { createEntityProxy } = require('./../../../_')
// Books
module.exports.Book = createEntityProxy(['sap.capire.bookshop', 'Books'], { target: { is_singular: true } })
module.exports.Books = createEntityProxy(['sap.capire.bookshop', 'Books'], { target: { is_singular: false }})
// Authors
module.exports.Author = createEntityProxy(['sap.capire.bookshop', 'Authors'], { target: { is_singular: true } })
module.exports.Authors = createEntityProxy(['sap.capire.bookshop', 'Authors'], { target: { is_singular: false }})
// Genres
module.exports.Genre = createEntityProxy(['sap.capire.bookshop', 'Genres'], { target: { is_singular: true } })
module.exports.Genres = createEntityProxy(['sap.capire.bookshop', 'Genres'], { target: { is_singular: false }})
// Books.texts
module.exports.Books.text = createEntityProxy(['sap.capire.bookshop', 'Books.texts'], { target: { is_singular: true } })
module.exports.Books.texts = createEntityProxy(['sap.capire.bookshop', 'Books.texts'], { target: { is_singular: false }})
// Genres.texts
module.exports.Genres.text = createEntityProxy(['sap.capire.bookshop', 'Genres.texts'], { target: { is_singular: true } })
module.exports.Genres.texts = createEntityProxy(['sap.capire.bookshop', 'Genres.texts'], { target: { is_singular: false }})
// events
// actions
// enums
