// @msggroup-softproviding/eslint-plugin/dist/rules/no-where-clause.js
import { ESLintUtils } from "@typescript-eslint/utils";
const createRule = ESLintUtils.RuleCreator((name) => `https://example.com/rule/${name}`);
export const requireWhereClause = createRule({
    name: "require-where-clause",
    meta: {
        type: "problem",
        docs: {
            description: "Require .where() method call after SELECT(...) and SELECT.from(...)",
        },
        fixable: undefined,
        schema: [],
        messages: {
            missingWhere: "SELECT queries must include a .where() clause for safety",
        },
    },
    defaultOptions: [],
    create(context) {
        return {
            CallExpression(node) {
                // Check for SELECT.from(...) calls
                if (node.callee.type === "MemberExpression" &&
                    node.callee.object.type === "Identifier" &&
                    node.callee.object.name === "SELECT" &&
                    node.callee.property.type === "Identifier" &&
                    node.callee.property.name === "from") {
                    checkForWhereClause(node, context);
                }
            },
        };
    },
});
function checkForWhereClause(node, context) {
    // Find the parent expression that contains the full method chain
    let currentNode = node;
    let hasWhere = false;
    // Traverse up to find the top-level expression
    while (currentNode.parent) {
        if (currentNode.parent.type === "MemberExpression") {
            currentNode = currentNode.parent;
        }
        else if (currentNode.parent.type === "CallExpression" &&
            currentNode.parent.callee === currentNode) {
            currentNode = currentNode.parent;
        }
        else {
            break;
        }
    }
    // Now traverse the method chain to look for .where()
    hasWhere = hasWhereInChain(currentNode);
    if (!hasWhere) {
        context.report({
            node: node,
            messageId: "missingWhere",
        });
    }
}
function hasWhereInChain(node) {
    if (node.type === "CallExpression") {
        // Check if this call is a .where() call
        if (node.callee.type === "MemberExpression" &&
            node.callee.property.type === "Identifier" &&
            node.callee.property.name === "where") {
            return true;
        }
        // Check the callee for chained calls
        if (node.callee.type === "MemberExpression") {
            return hasWhereInChain(node.callee.object);
        }
    }
    else if (node.type === "MemberExpression") {
        // Check if this member access is for .where
        if (node.property.type === "Identifier" && node.property.name === "where") {
            return true;
        }
        // Continue checking the object
        return hasWhereInChain(node.object);
    }
    return false;
}
// Export the rule
export default requireWhereClause;
