// @sap/cds-dk/lib/init/template/sample/files/@cds-models/AdminService/index.js
// This is an automatically generated file. Please do not change its contents manually!
const { createEntityProxy } = require('./../_')
// service
const AdminService = { name: 'AdminService' }
module.exports = AdminService
module.exports.AdminService = AdminService
// Authors
module.exports.Author = createEntityProxy(['AdminService', 'Authors'], { target: { is_singular: true } })
module.exports.Authors = createEntityProxy(['AdminService', 'Authors'], { target: { is_singular: false }})
// Books
module.exports.Book = createEntityProxy(['AdminService', 'Books'], { target: { is_singular: true } })
module.exports.Books = createEntityProxy(['AdminService', 'Books'], { target: { is_singular: false }})
// Genres
module.exports.Genre = createEntityProxy(['AdminService', 'Genres'], { target: { is_singular: true } })
module.exports.Genres = createEntityProxy(['AdminService', 'Genres'], { target: { is_singular: false }})
// Languages
module.exports.Language = createEntityProxy(['AdminService', 'Languages'], { target: { is_singular: true } })
module.exports.Languages = createEntityProxy(['AdminService', 'Languages'], { target: { is_singular: false }})
// Currencies
module.exports.Currency = createEntityProxy(['AdminService', 'Currencies'], { target: { is_singular: true } })
module.exports.Currencies = createEntityProxy(['AdminService', 'Currencies'], { target: { is_singular: false }})
// Books.texts
module.exports.Books.text = createEntityProxy(['AdminService', 'Books.texts'], { target: { is_singular: true } })
module.exports.Books.texts = createEntityProxy(['AdminService', 'Books.texts'], { target: { is_singular: false }})
// Genres.texts
module.exports.Genres.text = createEntityProxy(['AdminService', 'Genres.texts'], { target: { is_singular: true } })
module.exports.Genres.texts = createEntityProxy(['AdminService', 'Genres.texts'], { target: { is_singular: false }})
// Languages.texts
module.exports.Languages.text = createEntityProxy(['AdminService', 'Languages.texts'], { target: { is_singular: true } })
module.exports.Languages.texts = createEntityProxy(['AdminService', 'Languages.texts'], { target: { is_singular: false }})
// Currencies.texts
module.exports.Currencies.text = createEntityProxy(['AdminService', 'Currencies.texts'], { target: { is_singular: true } })
module.exports.Currencies.texts = createEntityProxy(['AdminService', 'Currencies.texts'], { target: { is_singular: false }})
// events
// actions
// enums
