// @sap/cds-dk/lib/init/template/sample/files/@cds-models/CatalogService/index.js
// This is an automatically generated file. Please do not change its contents manually!
const { createEntityProxy } = require('./../_')
// service
const CatalogService = { name: 'CatalogService' }
module.exports = CatalogService
module.exports.CatalogService = CatalogService
// ListOfBooks
module.exports.ListOfBook = createEntityProxy(['CatalogService', 'ListOfBooks'], { target: { is_singular: true } })
module.exports.ListOfBooks = createEntityProxy(['CatalogService', 'ListOfBooks'], { target: { is_singular: false }})
// Books
module.exports.Book = createEntityProxy(['CatalogService', 'Books'], { target: { is_singular: true } })
module.exports.Books = createEntityProxy(['CatalogService', 'Books'], { target: { is_singular: false }})
// Genres
module.exports.Genre = createEntityProxy(['CatalogService', 'Genres'], { target: { is_singular: true } })
module.exports.Genres = createEntityProxy(['CatalogService', 'Genres'], { target: { is_singular: false }})
// Currencies
module.exports.Currency = createEntityProxy(['CatalogService', 'Currencies'], { target: { is_singular: true } })
module.exports.Currencies = createEntityProxy(['CatalogService', 'Currencies'], { target: { is_singular: false }})
// Books.texts
module.exports.Books.text = createEntityProxy(['CatalogService', 'Books.texts'], { target: { is_singular: true } })
module.exports.Books.texts = createEntityProxy(['CatalogService', 'Books.texts'], { target: { is_singular: false }})
// Genres.texts
module.exports.Genres.text = createEntityProxy(['CatalogService', 'Genres.texts'], { target: { is_singular: true } })
module.exports.Genres.texts = createEntityProxy(['CatalogService', 'Genres.texts'], { target: { is_singular: false }})
// Currencies.texts
module.exports.Currencies.text = createEntityProxy(['CatalogService', 'Currencies.texts'], { target: { is_singular: true } })
module.exports.Currencies.texts = createEntityProxy(['CatalogService', 'Currencies.texts'], { target: { is_singular: false }})
// events
// actions
module.exports.submitOrder = 'submitOrder'
// enums
