// @sap/cds-mtxs/@cds-models/cds/xt/index.js
// This is an automatically generated file. Please do not change its contents manually!
const { createEntityProxy } = require('./../../_')
// Extensions
module.exports.Extension = createEntityProxy(['cds.xt', 'Extensions'], { target: { is_singular: true } })
module.exports.Extensions = createEntityProxy(['cds.xt', 'Extensions'], { target: { is_singular: false }})
// Tenants
module.exports.Tenant = createEntityProxy(['cds.xt', 'Tenants'], { target: { is_singular: true } })
module.exports.Tenants = createEntityProxy(['cds.xt', 'Tenants'], { target: { is_singular: false }})
// Jobs
module.exports.Job = createEntityProxy(['cds.xt', 'Jobs'], { target: { is_singular: true } })
module.exports.Jobs = createEntityProxy(['cds.xt', 'Jobs'], { target: { is_singular: false }})
// Tasks
module.exports.Task = createEntityProxy(['cds.xt', 'Tasks'], { target: { is_singular: true } })
module.exports.Tasks = createEntityProxy(['cds.xt', 'Tasks'], { target: { is_singular: false }})
// CSN
module.exports.CSN = createEntityProxy(['cds.xt', 'CSN'], { target: { is_singular: true } })
// I18N
module.exports.I18N = createEntityProxy(['cds.xt', 'I18N'], { target: { is_singular: true } })
// events
// actions
// enums
module.exports.Status ??= { QUEUED: "QUEUED", RUNNING: "RUNNING", FINISHED: "FINISHED", CANCELLED: "CANCELLED", TIMEOUT: "TIMEOUT", FAILED: "FAILED" }
module.exports.CSNConsumer ??= { nodejs: "nodejs", java: "java" }
