// @sap/cds-dk/lib/init/template/sample/files/@cds-models/CatalogService/index.ts
// This is an automatically generated file. Please do not change its contents manually!
import * as _sap_capire_bookshop from './../sap/capire/bookshop';
import * as __ from './../_';
import * as _ from './..';
import * as _sap_common from './../sap/common';

export default class {
  declare static readonly submitOrder: typeof submitOrder;
}

export function _ListOfBookAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class ListOfBook extends Base {
    declare createdAt?: __.CdsTimestamp | null
    declare modifiedAt?: __.CdsTimestamp | null
    declare ID?: __.Key<number>
    declare author?: string | null
    declare title?: string | null
    declare genre?: string | null
    declare stock?: number | null
    declare price?: _sap_capire_bookshop.Price | null
    declare currency?: string | null
    declare texts?: __.Composition.of.many<Books.texts>
    declare localized?: __.Association.to<Books.text> | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<ListOfBook>;
    declare static readonly elements: __.ElementsOf<ListOfBook>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
/** For displaying lists of Books */
export class ListOfBook extends _ListOfBookAspect(__.Entity) {}
Object.defineProperty(ListOfBook, 'name', { value: 'CatalogService.ListOfBooks' })
Object.defineProperty(ListOfBook, 'is_singular', { value: true })
/** For displaying lists of Books */
export class ListOfBooks extends Array<ListOfBook> {$count?: number}
Object.defineProperty(ListOfBooks, 'name', { value: 'CatalogService.ListOfBooks' })

export function _BookAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Book extends Base {
    declare createdAt?: __.CdsTimestamp | null
    declare modifiedAt?: __.CdsTimestamp | null
    declare ID?: __.Key<number>
    declare author?: string | null
    declare title?: string | null
    declare descr?: string | null
    declare genre?: __.Association.to<Genre> | null
    declare genre_ID?: string | null
    declare stock?: number | null
    declare price?: _sap_capire_bookshop.Price | null
    /**
    * Type for an association to Currencies
    * 
    * See https://cap.cloud.sap/docs/cds/common#type-currency
    */
    declare currency?: _.Currency | null
    declare currency_code?: string | null
    declare texts?: __.Composition.of.many<Books.texts>
    declare localized?: __.Association.to<Books.text> | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Book>;
    declare static readonly elements: __.ElementsOf<Book>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
/** For display in details pages */
export class Book extends _BookAspect(__.Entity) {}
Object.defineProperty(Book, 'name', { value: 'CatalogService.Books' })
Object.defineProperty(Book, 'is_singular', { value: true })
/** For display in details pages */
export class Books extends Array<Book> {$count?: number}
Object.defineProperty(Books, 'name', { value: 'CatalogService.Books' })

export function _GenreAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Genre extends Base {
    declare ID?: __.Key<string>
    declare name?: string | null
    declare descr?: string | null
    declare parent?: __.Association.to<Genre> | null
    declare parent_ID?: string | null
    declare children?: __.Composition.of.many<Genres>
    declare texts?: __.Composition.of.many<Genres.texts>
    declare localized?: __.Association.to<Genres.text> | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Genre>;
    declare static readonly elements: __.ElementsOf<Genre>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
/** Hierarchically organized Code List for Genres */
export class Genre extends _GenreAspect(__.Entity) {}
Object.defineProperty(Genre, 'name', { value: 'CatalogService.Genres' })
Object.defineProperty(Genre, 'is_singular', { value: true })
/** Hierarchically organized Code List for Genres */
export class Genres extends Array<Genre> {$count?: number}
Object.defineProperty(Genres, 'name', { value: 'CatalogService.Genres' })

export function _CurrencyAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
  return class Currency extends Base {
    declare name?: string | null
    declare descr?: string | null
    declare code?: __.Key<string>
    declare symbol?: string | null
    declare minorUnit?: number | null
    declare numcode?: number | null
    declare exponent?: number | null
    declare minor?: string | null
    declare texts?: __.Composition.of.many<Currencies.texts>
    declare localized?: __.Association.to<Currencies.text> | null
    static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
    declare static readonly keys: __.KeysOf<Currency>;
    declare static readonly elements: __.ElementsOf<Currency>;
    declare static readonly actions: globalThis.Record<never, never>;
  };
}
/**
* Code list for currencies
* 
* See https://cap.cloud.sap/docs/cds/common#entity-currencies
*/
export class Currency extends _CurrencyAspect(__.Entity) {}
Object.defineProperty(Currency, 'name', { value: 'CatalogService.Currencies' })
Object.defineProperty(Currency, 'is_singular', { value: true })
/**
* Code list for currencies
* 
* See https://cap.cloud.sap/docs/cds/common#entity-currencies
*/
export class Currencies extends Array<Currency> {$count?: number}
Object.defineProperty(Currencies, 'name', { value: 'CatalogService.Currencies' })


export declare const submitOrder:  {
  // positional
  (book: __.Key<__.DeepRequired<Book>['ID']>, quantity: number | null): globalThis.Promise< {
  stock?: number | null,
} | null> |  {
  stock?: number | null,
} | null
  // named
  ({book, quantity}: {book?: __.Key<__.DeepRequired<Book>['ID']>, quantity?: number | null}): globalThis.Promise< {
  stock?: number | null,
} | null> |  {
  stock?: number | null,
} | null
  // metadata (do not use)
  __parameters: {book?: __.Key<__.DeepRequired<Book>['ID']>, quantity?: number | null}, __returns: globalThis.Promise< {
  stock?: number | null,
} | null> |  {
  stock?: number | null,
} | null, __self: never
  kind: 'action'
}
export namespace Books {
  export function _textAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
    return class text extends Base {
      declare ID_texts?: __.Key<string>
      /** Type for a language code */
      declare locale?: _sap_common.Locale | null
      declare ID?: number | null
      declare title?: string | null
      declare descr?: string | null
      static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
      declare static readonly keys: __.KeysOf<text>;
      declare static readonly elements: __.ElementsOf<text>;
      declare static readonly actions: globalThis.Record<never, never>;
    };
  }
  export class text extends _textAspect(__.Entity) {}
  Object.defineProperty(text, 'name', { value: 'CatalogService.Books.texts' })
  Object.defineProperty(text, 'is_singular', { value: true })
  export class texts extends Array<text> {$count?: number}
  Object.defineProperty(texts, 'name', { value: 'CatalogService.Books.texts' })
  
}
export namespace Genres {
  export function _textAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
    return class text extends Base {
      /** Type for a language code */
      declare locale?: __.Key<_sap_common.Locale>
      declare ID?: __.Key<string>
      declare name?: string | null
      declare descr?: string | null
      static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
      declare static readonly keys: __.KeysOf<text>;
      declare static readonly elements: __.ElementsOf<text>;
      declare static readonly actions: globalThis.Record<never, never>;
    };
  }
  export class text extends _textAspect(__.Entity) {}
  Object.defineProperty(text, 'name', { value: 'CatalogService.Genres.texts' })
  Object.defineProperty(text, 'is_singular', { value: true })
  export class texts extends Array<text> {$count?: number}
  Object.defineProperty(texts, 'name', { value: 'CatalogService.Genres.texts' })
  
}
export namespace Currencies {
  export function _textAspect<TBase extends new (...args: any[]) => object>(Base: TBase) {
    return class text extends Base {
      /** Type for a language code */
      declare locale?: __.Key<_sap_common.Locale>
      declare name?: string | null
      declare descr?: string | null
      declare code?: __.Key<string>
      static readonly kind: 'entity' | 'type' | 'aspect' = 'entity';
      declare static readonly keys: __.KeysOf<text>;
      declare static readonly elements: __.ElementsOf<text>;
      declare static readonly actions: globalThis.Record<never, never>;
    };
  }
  export class text extends _textAspect(__.Entity) {}
  Object.defineProperty(text, 'name', { value: 'CatalogService.Currencies.texts' })
  Object.defineProperty(text, 'is_singular', { value: true })
  export class texts extends Array<text> {$count?: number}
  Object.defineProperty(texts, 'name', { value: 'CatalogService.Currencies.texts' })
  
}
